src/nvim/api/vim.c

```diff
@@ -164,6 +164,83 @@ Object vim_eval(String str, Error *err)
   return rv;
 }
 
+/// Call the given function with the given arguments stored in an array.
+///
+/// @param functionName The function to call
+/// @param args the functions arguments packed in an Array
+/// @param self optional self dictionary that may be used by user functions to
+///        emulate object-oriented programming
+/// @param[out] err Details of an error that may have occurred
+/// @return The result of the function call if it was successful
+Object vim_call_function(String functionName, Array args, Object self,
+                         Error *err)
+  FUNC_ATTR_DEFERRED
+{
+  Object rv = OBJECT_INIT;
+  if (args.size > MAX_FUNC_ARGS) {
+    api_set_error(err, Validation,
+      _("Function called with too many arguments."));
+    return rv;
+  }
+
+  typval_T optional_self_dict;
+  dict_T *self_dict = NULL;
+  if (self.type == kObjectTypeNil) {
+    self_dict = NULL;
+  } else if (self.type == kObjectTypeDictionary
+             || self.type == kObjectTypeArray) {
+    // XXX The tests somehow cause the object to be an array
+    self.type = kObjectTypeDictionary;
+    if (!object_to_vim(self, &optional_self_dict, err)) {
+      free_tv(&optional_self_dict);
+      return rv;
+    } else if (optional_self_dict.v_type == VAR_DICT) {
+      self_dict = optional_self_dict.vval.v_dict;
+    } else {
+      api_set_error(err, Validation,
+          _("Object was not converted to a dict."));
+      goto free_vim_args;
+    }
+  } else {
```

Comment:
Should we also handle the case where `self` might be the name of global dictionary? Using message pack to define a dictionary via an argument seems obtuse to me.